YAML front matter parsing

roborourke opened the pull request:
Fixes #1

roborourke committed:
first pass YAML front matter parsing

rmccue commented on inc/class-importer.php:
```diff
@@ -281,11 +281,20 @@ protected function update_post_from_markdown_source( $post_id ) {
+			$yaml_parser = new Yaml();
+			$yaml        = $yaml_parser->loadString( $yaml[1] );
+			// Strip YAML doc from the header
+			$markdown = preg_replace( '#^---(.+)---#Us', '', $markdown );
```
This could use `substr( $markdown, strlen( $yaml[0] ) )` (I'd rename `$yaml` to `$yaml_matches` though).

rmccue commented on inc/class-importer.php:
```diff
@@ -313,9 +322,24 @@ protected function update_post_from_markdown_source( $post_id ) {
+		// Add meta data from YAML front matter.
+		if ( isset( $yaml['meta'] ) && is_array( $yaml['meta'] ) ) {
+			foreach ( $yaml['meta'] as $key => $value ) {
+				update_post_meta( $post_id, $key, $value );
```
`$key` and `$value` both need to be slashed. Also, might be worth using a filterable whitelist of keys here.

roborourke committed:
added other relevant YAML front matter features

roborourke committed:
remove redundant guard

roborourke commented:
@rmccue ready for reviewing again. Added support for a few other bits of front matter.

roborourke committed:
namespace actions more

roborourke committed:
fix regex for title if newlines before it

roborourke committed:
fix regex for title if newlines before it

roborourke committed:
add author support

rmccue commented on inc/class-importer.php:
```diff
@@ -311,11 +322,51 @@ protected function update_post_from_markdown_source( $post_id ) {
+			wp_set_post_terms( $post_id, wp_slash( $yaml['category'] ), 'category' );
+		}
+		if ( isset( $yaml['categories'] ) && is_object_in_taxonomy( $this->get_post_type(), 'category' ) ) {
+			wp_set_post_terms( $post_id, array_map( 'wp_slash', $yaml['categories'] ), 'category' );
```
Is there a reason we need both `category` and `categories` (apart from Jekyll compat, which I'm not too concerned about)?

rmccue commented on inc/class-importer.php:
```diff
@@ -311,11 +322,51 @@ protected function update_post_from_markdown_source( $post_id ) {
 
+		// Add meta data from YAML front matter.
+		if ( isset( $yaml['meta'] ) && is_array( $yaml['meta'] ) ) {
+			$whitelist_keys = apply_filters( 'wordpressdotorg.markdown_sync.meta_whitelist', array(
```
This needs phpDoc.

rmccue reviewed (changes requested):
This also needs documentation of the supported keys and values, either in README.md or in a separate doc file.

rmccue commented:
cc @danielbachhuber for feedback here too

roborourke committed:
add phpdoc, remove category support

danielbachhuber commented on inc/class-importer.php:
```diff
@@ -311,11 +322,71 @@ protected function update_post_from_markdown_source( $post_id ) {
+		if ( isset( $yaml['date'] ) ) {
+			$post_data['post_date'] = date( 'Y-m-d H:i:s', strtotime( $yaml['date'] ) );
+		}
+		if ( isset( $yaml['author'] ) ) {
```
What about securities?

roborourke replied:
I guess it wouldn't really change things that much - if a user is found they would still only be able to make changes if their user role for the site allowed it. Not too familiar with how things are set up on wp.org.

roborourke replied:
Given changes go through a review cycle and that it's totally optional to add that I don't think it's a big problem personally.

danielbachhuber commented on inc/class-importer.php:
```diff
@@ -311,11 +322,71 @@ protected function update_post_from_markdown_source( $post_id ) {
+			 * @param array $whitelist_keys The meta keys to be imported if found.
+			 * @param array $yaml           The parsed YAML front matter.
+			 */
+			$whitelisted_keys = apply_filters( 'wordpressdotorg.markdown_sync.meta_whitelist', array(
```
😛

danielbachhuber commented:
Generally looks fine on my end. It's been quite a while since I touched my original code so I don't much remember its strengths and weaknesses.

Re: security comment, I guess we need to make sure these docs repos generally follow permissions for the WordPress site (e.g. anyone that has merge access in the docs repo probably should be an editor on the site, for visibility)

danielbachhuber commented:
Also, one thing that would be nice is some endpoint to trigger the import that GitHub can ping with push events.

roborourke commented:
Agree regarding the webhook but that's one for another issue, and I think for the dotorg site repo where the importer is being extended.

tcrsavage committed:
Use wp_set_object_terms for yaml categories

Bypasses the is_int check in wp_set_post_terms which stops assignement of cats via slugs for hierarchical taxonomies.

roborourke committed:
change to title logic, check YAML for excerpt too

roborourke committed:
Merge branch 'yaml-parser' of github.com:humanmade/markdown-sync into yaml-parser

roborourke committed:
fix regex for stripping title. Got borked at some point!

roborourke committed:
dont strip excerpts from content

roborourke committed:
excertp is up to first double newline, fix title replacement regex

rmccue commented:
@roborourke Is this ready to go?

roborourke commented:
@rmccue yeah I think so, being used in anger on an internal project and working well so far

rmccue committed:
Merge branch 'master' into yaml-parser

rmccue committed:
Add post_data filter for efficiency

rmccue committed:
Move description check to other excerpt code

rmccue committed:
Remove new fields by default

These can be added back in via a filter instead.

rmccue committed:
Remove all default meta fields for YAML

This can be handled by filters instead.

rmccue committed:
Slash all post data at once

rmccue commented:
I removed all the new fields, but left `title` and `description` overrides. Also added a new filter to allow for updating additional post fields at the same time. This should allow us to safely merge for insecure contexts (i.e. WordPress.org) while allowing flexibility for internal/private repos.

rmccue committed:
Avoid bonus slashes

fklein-lu reviewed (approved):
A small improvement would be to work on the return value of the `update_post_from_markdown_source()` by checking the return values of the database updating functions.

Right now the function returns `true`, even if something went wrong during the update.

rmccue commented:
Agreed, but was going to handle that in a separate PR. Thanks for the review! ✨
